RavenDB-15020 : Additional time series functions

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-15020

aviv86 committed:
RavenDB-15020 : Additional time series functions - initial work on 'percentile'

aviv86 committed:
RavenDB-15020 : time series query - add support for computing percentiles

aviv86 committed:
RavenDB-15020 : initial work on slope

aviv86 committed:
RavenDB-15641 : TimeSeriesAggregation - avoid computing final values more than once

aviv86 committed:
RavenDB-15020 : fixes, cleanup, tests

karmeli87 commented on src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -145,7 +145,7 @@ public class TimeSeriesAggregationResult<T> : TimeSeriesQueryStreamResultBase<Ti
     {
         public long[] Count;
-        public double[] Max, Min, Last, First, Average, Sum;
+        public double[] Max, Min, Last, First, Average, Sum, Percentile, Slope;
```
Wonder if we should call it `Derivative` (or maybe `Change`?) instead of `Slope`.

ayende replied:
Isn't slope the right math term?

aviv86 replied:
I think `Slope` is the right math term. 
`Change` is inaccurate - we are measuring **the rate** of change
and `Derivative` is the rate of change when the distance (delta X) tends towards 0, no?

karmeli87 commented on src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -0,0 +1,414 @@
+                    case AggregationType.Percentile:
+                        Debug.Assert(_rankedValues != null);
+                        var currentRankedList = _rankedValues[i];
+                        var loc = currentRankedList.BinarySearch(val);
```
I would prefer to use `SortedDictionary` with a comparer to handle duplicates.

ayende replied:
Agreed, also, and `insert` here will cause us to move a lot of values, potentially.

aviv86 replied:
done

aviv86 committed:
RavenDB-15020 : TimeSeriesAggregation refactoring

ayende commented on src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -0,0 +1,414 @@
+
+        private IEnumerable<double> GetPercentile(double scale)
+        {
+            foreach (var list in _rankedValues)
```
This rounds down, right? Do we need some manner to adjust the value based on the distance?
So if we have [1,2,4] and we ask for 75% percentile, we'll get [2] here. But that actually represent the 50%.
Should we adjust that so it would be 3 as the result?

aviv86 replied:
No actually we round up : `index = (int)Math.Ceiling(_percentileFactor * list.Count)`
so in your example we'll get `4` as the result. (`2` is wrong by any definition)

There are different definitions for percentile and methods for calculating percentile.
By the definition that I chose to implement the result is always a part of the data set.
So if we want to adjust the result to be `3` we need to use a different method..

I wrote more about it in the issue page : https://issues.hibernatingrhinos.com/issue/RavenDB-15020#focus=Comments-67-51024.0-0

ayende replied:
Okay, that make sense.
If it isn't hard, let's go with extrapolation. Note that we need to document this.

aviv86 replied:
done

aviv86 committed:
RavenDB-15020 : address PR comment - use sorted dictionary in PercentileAggregation

aviv86 committed:
RavenDB-15020 : Additional time series functions - add support for standard deviation

aviv86 committed:
RavenDB-15020 : fixes, cleanup

aviv86 committed:
RavenDB-15020 : PercentileAggregation : change implementation to use linear interpolation between closest ranks method

aviv86 commented:
- Added `StandardDeviation` support
- Made some refactoring in `TimeSeriesAggregation`  (introduced `ITimeSeriesAggregation`)
- Addressed PR comments

ppekrol commented on src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        IEnumerable<double> ITimeSeriesAggregation.GetFinalValues(DateTime? @from, DateTime? to, double? scale = null)
+        {
```
Can you check the warning?
```
src/Raven.Server/Documents/Queries/Results/TimeSeries/AverageAggregation.cs#L83
The default value specified for parameter 'scale' will have no effect because it applies to a member that is used in contexts that do not allow optional arguments
```

Probably just remove `= null` from `scale` here. It is an explicit interface impl.

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs:
```diff
@@ -0,0 +1,144 @@
+            }
+        }
+
+        IEnumerable<double> ITimeSeriesAggregation.GetFinalValues(DateTime? @from, DateTime? to, double? scale = null)
```
Warning here. Explicit interface impl.

aviv86 replied:
done

ppekrol commented on src/Raven.Server/Documents/Queries/Results/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -0,0 +1,274 @@
+
+    }
+
+    public class TimeSeriesAggregationBase
```
abstract?

aviv86 replied:
done

aviv86 committed:
RavenDB-15020 : address PR comments

karmeli87 commented on src/Raven.Client/Documents/Queries/TimeSeries/TimeSeriesAggregation.cs:
```diff
@@ -145,7 +145,7 @@ public class TimeSeriesAggregationResult<T> : TimeSeriesQueryStreamResultBase<Ti
     {
         public long[] Count;
-        public double[] Max, Min, Last, First, Average, Sum;
+        public double[] Max, Min, Last, First, Average, Sum, Percentile, Slope, StandardDeviation;
```
Should we maybe shorten this to StdDev?

ayende replied:
yes

ppekrol replied:
No, then Average should be Avg, etc. Either use full names of abbreviations. Not both.

ppekrol replied:
Actually it is already mixed, nvm.

ml054 replied:
May I ask what's definition used for slope?

aviv86 replied:
The rate of change within a range (only works with group by)
i.e. `slope = dY/dX`, dY = `range.LastVal - range.FirstVal`, dX = `(range.End - range.Start).TotalMilliseconds`

ml054 replied:
ok I thought  it uses linear regression

karmeli87 commented on src/Raven.Server/Documents/Queries/Results/TimeSeries/SlopeAggregation.cs:
```diff
@@ -0,0 +1,144 @@
+            {
+                var val = values[i];
+
+                if (i < oldCount)
```
for consistency with above `oldCount <= i`

karmeli87 reviewed (approved):
Very Nice! I like the `TimeSeriesAggregation` abstraction :-)
